wrappers/mongo/mongo.go

```diff
@@ -0,0 +1,641 @@
+
+package epsagonmongo
+
+import (
+	"context"
+	"fmt"
+	"github.com/epsagon/epsagon-go/tracer"
+	"reflect"
+
+	"github.com/epsagon/epsagon-go/epsagon"
+	"go.mongodb.org/mongo-driver/mongo"
+	mongoOptions "go.mongodb.org/mongo-driver/mongo/options"
+)
+
+
+// MongoCollectionWrapper is Epsagon's wrapper for mongo.Collection
+type MongoCollectionWrapper struct {
+	collection		*mongo.Collection
+	tracer 	tracer.Tracer
+}
+
+func WrapMongoCollection(
+	collection *mongo.Collection, ctx ...context.Context,
+) *MongoCollectionWrapper {
+	return &MongoCollectionWrapper{
+		collection:	collection,
+		tracer: 	epsagon.ExtractTracer(ctx),
+	}
+}
+
+
+func (coll *MongoCollectionWrapper) Name() string {
+	return coll.collection.Name()
+}
+
+func (coll *MongoCollectionWrapper) Database() *mongo.Database {
+	return coll.collection.Database()
+}
+
+func (coll *MongoCollectionWrapper) Clone(opts ...*mongoOptions.CollectionOptions) (interface{}, error) {
+	event := startMongoEvent(currentFuncName(), coll)
+	response, err := coll.collection.Clone(
+		opts...,
+	)
+	if err != nil {
+		logOperationFailure(fmt.Sprintf("Could not complete %s", currentFuncName()), err.Error())
+		epsagon.ExtractTracer([]context.Context{}).AddExceptionTypeAndMessage(
+			"mongo-driver",
+			err.Error(),
+		)
+	}
+	completeMongoEvent(event)
+
+	return response, err
+}
+
+
+func (coll *MongoCollectionWrapper) InsertOne(
+	ctx context.Context, document interface{}, opts ...*mongoOptions.InsertOneOptions,
+) (*mongo.InsertOneResult, error) {
+	defer epsagon.GeneralEpsagonRecover("mongo-driver", currentFuncName(), coll.tracer)
```

Comment:
actually when thinking about it, should use the GeneralEpsagonRecover only in internal functions - startMongoEvent, and maybe in completeMongoEvent (if you think it might fail). When used in startMongoEvent, it might return a nil pointer instead of an event - and you should detect it and not call complete event data. Let's talk about it and I will explain exactly what I mean

Reply:
moved to only internal